### ISSUE TITLE
A try at improving how the map layers are displayed.

### ISSUE THREAD

**Issue by Bertram25** [ValyriaTear/ValyriaTear#435]
@authenticate I must say it is more a question than a feature desired in game, but since you have the technical background about this, I thought I would just ask out. ;)

At the moment, the map layers are drawn tiles per tiles for each layers. I've nothing about it but since we're drawing them in a magnified way, the tile grid is unfortunately visible even on tile with the same color, due to the way OpenGL magnify textures.

![vt-diff-tileset](https://cloud.githubusercontent.com/assets/765917/6151587/26781362-b217-11e4-9a3e-fa5106b45d51.png)

To fix this, I wondered whether we could draw the map layers to temporary texture at their actual size, thus avoiding to magnify them and then magnify the whole layer when drawing it.
This could also permit us to render tiles per tiles only once and then render part of the whole layer depending on the map offset.

This would also later permit us to add random zoom support.

What do you think of the idea? Does this sound a sane way to do it using OpenGL?

**Comment by authenticate** [comment]
If possible, could you upload a screen shot of the problem?  I think I understand what you are describing, however, it doesn't hurt to make sure. :)

My initial answer is I am not entirely sure.

So, I would recommend making a simple test case and loading it into the game engine.

For example, I would create something like the following in the game:

![squares](https://cloud.githubusercontent.com/assets/4543259/6134262/ff8f6104-b12d-11e4-8d0c-e762811f0354.png)

(All three squares are supposed to be the same size.  My drawing is bad. :))

Let's assume tiles "A" and "B" are designed to be drawn next to each other on a map.  I would make tile "C", using Photoshop or GIMP, such that the boundary between "A" and "B" is in the middle of "C".  Then, I would use the editor to place tile "C" on a map near "A" and "B".  This way you can run the game and see exactly what would happen.

Now, if this fixes the problem, I think the solution is good.  However, there may be a few technical hurdles.  OpenGL does have a maximum texture size.  (I think it varies with the hardware on a machine.)  So, it seems possible that our larger maps, if aggregated into a single map texture, could be larger than that size.



In regards to zooming, I don't think making the map into a single texture really affects a zooming implementation.

Let's assume the map zooming works.  How do we render the entities on the map?  What are their positions relative to the "zoomed" map?  I think solving that problem is the same problem, or at least a very similar problem, to zooming the map per tile.

I have to think about this some more.

Zooming in 3D is easy.  The math for a 3D game is well defined and engineered to "just work".  Now, this does make the math a little more complicated.  However, that's the trade off.

In most 2D games, people take short cuts to simplify the math, which is fine.  However, zooming is one of the operations that sort of breaks down if you simplify things.  It ends up usually being harder in a 2D game.

In VT, we use a 3D transformation stack.  It is just set up and configured for a 2D game.  So, we theoretically should already have zooming 90% implemented.  I would have to look through the code and experiment with some things to know for sure though.  I could be wrong here.

**Comment by rootslinux** [comment]
Regarding zooming, you also need to consider that the 2D graphics are constructed with pixel art. If you want to gradually move the zoom from 1x to 2x, for example, you need to choose an appropriate scaling algorithm. If you've ever used an emulator for an old console (SNES, for example), in the options you'll see that you can choose from among several different scaling algorithms.

http://en.wikipedia.org/wiki/Image_scaling

I've been considering doing something similar for Allacrost maps in the future, being able to go from our current map size (where all images are zoomed to 2x) to a native map size (images drawn at 1x zoom). Haven't really looked into this at all other than thinking about it though.

**Comment by Bertram25** [comment]
@authenticate  Ok, I'll upload a more precise example asap.
As for the zooming part, I must say I wasn't even on that part actually, but rather thinking about keeping this in mind while we're on the design part.

Like you @rootslinux I've done nothing more than thinking about it. :) But here, I wasn't speaking about using fixed zooming + some 2D art refiner algorithm but more dynamic zooms, just like for special actions in battles, and such.

**Comment by Bertram25** [comment]
Here is a visual example of the bug I spoke about, and IMHO you guessed it well @authenticate  
![vt-diff-tileset](https://cloud.githubusercontent.com/assets/765917/6151587/26781362-b217-11e4-9a3e-fa5106b45d51.png)
To the left, on the magnified part of the image, you should be able to see the grid where the tiles have faded because of being drawn one by one.
On the right, everything is fine. Would there be a bug on the right, it would then come from the tileset.

As you pointed out, drawing the whole layer on a texture comes with problems when it comes to texture max size, but maybe could we draw each visible layer part on 1024x768 (1024x1024) textures and render those?
We could even update them only if the map scrolling offset changes, earning a few fps along the way?

WDYT?

**Comment by authenticate** [comment]
I did some experimenting and I think the core issue is using linear filtering (GL_LINEAR) for rendering the scaled map tiles.

For example, I arbitrarily changed all the filtering to nearest neighbor (GL_NEAREST) and the map no longer had those visual anomalies.  All the other parts of the game, like text rendering, looked bad.  However, if there is an easy way just to render the map with nearest neighbor and the remainder of the game linearly, that might be the quickest and easiest solution.

Your solution now makes a lot of sense to me.  Linear filtering struggles on boundaries since it does not have accurate data to interpolate from the other side of the boundary.  By compositing the image first at the original size and then scaling it, you eliminate the boundaries.

As an additional approach, you could implement what you described using multiple render targets.

We could allocate a secondary texture which is also a render target.  For simplicity, it could be the same dimensions as the original render target which gets displayed on the screen.

Then, we could render the map at its original size into the secondary render target.  Now, what position would we render the map relative to this render target?  That's a non-trivial question.  I am not really sure.

However, once we have the secondary render target composited, we can render a scaled subset of the secondary render target onto the original render target.  This would draw the entire visible map without the filtering problems.  However, like before, I feel it might be deceptively difficult.  How exactly would we calculate this subset?

This is just something to think about.  If you already have the code and infrastructure to composite images on the CPU, it would probably be easier and safer to use that.  However, I think you will run into the same type of position and offset questions.

This whole topic essentially goes back to what Roots brought up.  We are scaling the map; our scaling algorithm does not do exactly would we would like it to do.  So, what do we do? :)

Investigating those algorithms might actually be the best use of time.  They were created exactly for scaling pixel art; which is our problem.

**Comment by rootslinux** [comment]
I spent a long time working on the map code a few years ago trying to get the "uglyness" of the tiles to go away. I knew that the solution lied somewhere in the graphics engine. So, if we change the filtering to GL_NEAREST all those artifacts on the map go away? I think the best solution is to do just as you said: use GL_NEAREST filtering when drawing map tiles, and GL_LINEAR everywhere else. That should be a relatively easy change to make to the graphics engine and will really improve the graphics quality of maps.

I'm still not sure what benefit there is to rendering the map to a secondary texture. What does it buy us? Is there a simple pros/cons list we can put together to figure out if pursuing this is worth it (especially for people like me who have no graphics background)?

As for the scaling, I think you just need to do some experimentation to figure out what looks/works best. But first we should definitely fix the glitches in the rendered map before playing around with that.

**Comment by rootslinux** [comment]
Took a look at the code tonight and by default the video engine has the _smooth_textures member set to true. When this is enabled, it serves as a global setting to set GL_LINEAR for all textures. Furthermore, the TexSheet  has a Smooth(bool flag = true) method that is called whenever the texture is loaded or reloaded.  When the flag is set to true, it uses GL_LINEAR, and uses GL_NEAREST if it is false. There's no real exposure in the video engine to change the filtering type of a texture.

I think the solution to do here would be:
1) Set the_smooth_textures member to false in the VideoEngine. In fact, maybe get rid of it altogether. All this does is serve to override the filters on all loaded textures, and I don't think it's useful to have.

2) In the LoadImage functions, allow an optional flag to be sent to specify when to use a different filter for the images to be loaded. Set a default value to use GL_LINEAR. Any tilesets that are loaded should make use of this flag so that the texture sheets the tiles are placed into use GL_NEAREST.


Slightly trickier than I thought it would be since the filter settings are all pretty much done internal to the engine. I'm going to play around and see if I can see the same results that authenticate did when he used GL_NEAREST. BTW, I looked at the Allacrost code when I investigated this, so I'm unsure if this is still valid for the VT code (pretty sure it is though, unless you made big changes to the video engine that I'm unaware of)

**Comment by authenticate** [comment]
I think it is very important that everyone is aware of the results before investing lots of time in a particular solution.

I would not describe switching to GL_NEAREST for the map as a perfect solution.  It's more like a trade-off.  It does help reduce the original problem described here.  However, it causes the map to look different which may not be desirable.  It depends on the game and the art style you are trying to achieve.

So, here is an example:

Linear (VT rendered as is)

![vt_linear](https://cloud.githubusercontent.com/assets/4543259/6172620/ff62d7a0-b2af-11e4-901c-bf026ee682a8.png)

Nearest

![vt_point](https://cloud.githubusercontent.com/assets/4543259/6172627/05f5f610-b2b0-11e4-983d-3482e611ba51.png)

(For nearest, I searched all files for GL_TEXTURE_MIN_FILTER and GL_TEXTURE_MAG_FILTER and changed the final parameter to GL_NEAREST; basically just hacked it in to test.)

(Note: You almost have to view the screenshots at their full size to really see the difference.)

So, I will let both of you decide if it's better, worse, or about the same visually as a whole.  Note: You have to pretend the characters and clouds in the nearest screen shot are actually rendered with GL_LINEAR.  That's sort of out of scope for the test.

Finally, I think it's also important to point out this process only works because the map is scaled to a power of two.  If you wanted to scale the map to three or to some irrational floating point number, the results would not look very good.

I'll get back to you on the questions.  I'll make a couple diagrams to help illustrate the issue.

**Comment by Bertram25** [comment]
As a first note, before, using GL_NEAREST or GL_LINEAR was a game option, hence the existence of the _smooth_textures member and related parameters in other functions. (Just have a look at the HEIR version in the boot options.)
This was indeed existing in Allacrost as well, and I made it so the algorithm does always use GL_LINEAR since I do think the game looks uglier without it, even when speaking only about the tiles.

So I'd still be for rendering the map layer part unscaled on some textures and re-render it scaled on screen, if it's a valid solution to fix that issue, and I would keep using GL_LINEAR for sprites and related visuals, since it's what's best looking so far.

Now, I'm aware the current engine hasn't been designed with dynamic zooming in mind from the start. (I don't blame anyone, of course, be sure of that.) but I still wonder whether we could aim toward that, even step per step, fixing the apparent grid due to linear scaling of tiles being the first one.

As a second step, and after thinking about it for a while, there may indeed nice algorithms to somewhat improve the 2D quality of the image, (like HQX or even 4xBRZ), or at least smooth it, but I would make it an option, applied on the final rendered target only, not on certain parts of it, and not applied on each part individually. And all that is an optional second step that I would make only if the first one isn't sufficient or is inapplicable.

In any case, thanks for all feed-back about it. It's cool and important to get different point of view about this before we consider making a move (or not.) :)

**Comment by Bertram25** [comment]
@authenticate 
> Your solution now makes a lot of sense to me. Linear filtering struggles on boundaries since it does not have accurate data to interpolate from the other side of the boundary. By compositing the image first at the original size and then scaling it, you eliminate the boundaries.
As an additional approach, you could implement what you described using multiple render targets.
We could allocate a secondary texture which is also a render target. For simplicity, it could be the same dimensions as the original render target which gets displayed on the screen.

I'm now at the point where I've cleaned up what I first wanted to cleanup in the new_map_code branch.
Before moving further, I'd like to add support so that both ground and sky map layers are rendered unscaled on a texture, and then simply drawn scaled using the map zoom ratio on screen.
See the new DrawLayer() function here: https://github.com/Bertram25/ValyriaTear/blob/new_map_code/src/modes/map/map_tiles.cpp#L326
I would tweak this function to draw the tiles unscaled to the texture.

Now, what is the best way to do that with the gl API at our disposal? Should I copy pixels on an array and load that array on an existing texture, or is there a better tweak, since tiles are normally already loaded in memory when being rendered?

Best regards,

**Comment by Bertram25** [comment]
While looking at rendering on a different target aka a frame buffer object, shouldn't the shaders be adapted for it or not?
I'm having a bit of fun at rendering the tiles on a generated frame buffer in map_tiles.cpp with direct gl calls atm (even if it still isn't ok at all) and I wondered whether there is some hint I'm missing.

**Comment by authenticate** [comment]
I think the fragment shaders have keywords to describe what information gets put where.  I have to double check this.  So, I think you will need to tell it to write a 4D vector (representing color) into the zero slot of the framebuffer object.  This would require creating a new shader program with a new fragment shader.  You should be able to reuse the default vertex shader in the shader program and use the sprite fragment shader as a starting point.

If you have a branch or some commits somewhere, I can take a look at it and help you out if you get stuck.

It's good news that the framebuffer objects are working initially though. :)

**Comment by Bertram25** [comment]
Well, I'll be honest, I can grasp the theory (and I'm at least happy about it) but I seriously lack practice with shaders, even if I understand they're just here to emulate FFP for now.

I'll have a go at following your advice and will show you what I've done. Time to take a good gl bath. ;)

**Comment by Bertram25** [comment]
So you know, I'll need more time for this one. I'm neck-deep busy and wanted on other projects atm (and still seeking a job, btw).

I'll try to get rid of other easier tasks little by little and get back to this one once I've got a few contiguous days free for that one.

**Comment by authenticate** [comment]
I found an article on the internet the other day related to this topic:

http://www.fortressofdoors.com/doing-an-hd-remake-the-right-way/

The relevant section is "Tile Artifacts".

I think we are approaching the problem correctly.  The article says to compose the scene first at its native resolution and then scale it.  This is the same approach we decided to implement; as long as I am understanding everything correctly.

**Comment by Bertram25** [comment]
Yeah, I spent something like two hours reading it yesterday. This article is an absolute jewel. :)
I even have a pass on the best ways to upscale graphics.

> I think we are approaching the problem correctly. The article says to compose the scene first at its native resolution and then scale it. This is the same approach we decided to implement; as long as I am understanding everything correctly.

Overall, it also makes me think it's the right way to go, indeed.

**Comment by Bertram25** [comment]
This was improvement thanks to off-screen rendering support made by @authenticate in https://github.com/ValyriaTear/ValyriaTear/pull/506

**Comment by Bertram25** [comment]
Well, I made the code to draw everything at real size and made the render be zoomed 2x to get the same result than before.
But then the overall results is even more blurry than what we currently have. So I guess we have a good compromise for now. I'll disable the second rendering target for maps for now to earn a few fps, and call this one finally out.
